src/typings.ts

```diff
@@ -146,41 +146,41 @@ interface LockRegistry {
 }
 
 interface DeepstreamConfig {
-  showLogo: boolean
-  libDir: string | null
-  logLevel: number
-  serverName: string
-  externalUrl: string | null
-  sslKey: string | null
-  sslCert: string | null
-  sslCa: string | null
-  connectionEndpoints: any
-
-  plugins: {
-    cache: PluginConfig
-    storage: PluginConfig
+  showLogo?: boolean
```

Comment:
The problem is internally we use a deepstreamconfig where non of these are optional, and a user can pass in a config object where they all are because they get merged in from default options.

Because of that I would argue (poorly) that we would need to have an external config object for typescript and an internal one rather than making the codebase have to check if it exists everywhere.

Reply:
Hi, I know what you mean, but using it with a type definitions mean that you must provide a full configuration object. Otherwise TypeScript will throw an error. The assertions, that all these properties exists is program logic and should be handled like this. And there are the checks, that these values exists. 

A solution could be to provide two interfaces and a check and transforming function. Typescript will notice in it, that you handle the constraint of existence and then it will accept the return type, where all properties are well set. This could be combined with an [Notification](https://martinfowler.com/eaaDev/Notification.html) pattern, M. Fowler described.

A quick example how a Transformation could work:

```typescript
interface ConfigurationOptional {
  value1?: number
  value2?: string
}

interface ConfigurationForce {
  value1: number
  value2: string
}

const assertType = <T>(defaultValue: T, givenValue?: T): T => {
  return typeof givenValue !== 'undefined' ? givenValue : defaultValue;
}

const defaultObject: ConfigurationForce = {
  value1: 2,
  value2: '3'
}

const assertConfiguration = (a: ConfigurationOptional): ConfigurationForce => {
  return {
    value1: assertType(defaultObject.value1, a.value1),
    value2: assertType(defaultObject.value2, a.value2)
  }
}
```
As you can see, you are able to transform from an optional to a forced constraint object. It is much more readable and you have the checks where you need them. You also have the opportunity to replace the merge method by subsequencing objects. It provide a fully checked object given by design. There have to be some rewrites but not so much. 

I hope it will help a little bit. If there are any questions, feel free to ask.

Reply:
Agreed, which is what I meant for:

> Because of that I would argue (poorly) that we would need to have an external config object for typescript and an internal one rather than making the codebase have to check if it exists everywhere.

The external config option will allow everything to be optional, the internal one will be the one you currently modified (as it is used everywhere).

The assertConfiguration method is already done within the configInitialiser, just need to return a different type.

Reply:
Hi,

the actual problem I see is, that you only implicit set values to an object but cannot predict that it happened. This makes it painful to analyze what happened, when something went wrong. This could happen by refactoring something in the modules or forget something to implement and so on. So we are not able to predict, that a function, which we call do what we expect from then todo.
A quick example:
```typescript
// bad
const configLoader = (cfg: any, defaultCfg: any) => {
    doSomething1(cfg, defaultCfg);
    doSomething2(cfg, defaultCfg);
    doSomething3(cfg, defaultCfg);
    doSomething4(cfg, defaultCfg);
    return cfg;
}

// good
const configLoaderGoodWay = (cfg: any, defaultCfg: any) => {
    if (!cfg) {
        return defaultCfg;
    }
    return {
        key1: doSomething1(cfg.key1, defaultCfg),
        key2: doSomething2(cfg.key2, defaultCfg),
        key3: doSomething3(cfg.key3, defaultCfg),
        key4: doSomething4(cfg.key4, defaultCfg),
    }
}
```
You can argue, that both of them do the same, but you cannot verify that there are doing. This is what TypeScript wants to force. If you have an optional value, than create an new object with another type of interface. That helps also to avoid side effects, I which are hard to check. 

And as you see, give your `doSomething` methods only the information they need to work.

If you wish I could make a few quick refactorings, to the problems I see. They will not break the the functionality, I think. So let me know what you are thinking about.